data_psi.go

```diff
@@ -449,3 +450,162 @@ func (d *PSIData) toData(firstPacket *Packet, pid uint16) (ds []*Data) {
 	}
 	return
 }
+
+func (d *PSIData) Serialise(b []byte) (int, error) {
+
+	//TODO take care of pointer field
+	if d.PointerField != 0 {
+		return 0, errors.New("Error pointer field muxing unimplemented")
+	}
+	b[0] = uint8(d.PointerField)
+	idx := 1
+	for i := range d.Sections {
+		n, err := d.Sections[i].Serialise(b[idx:])
+		if err != nil {
+			return idx, err
+		}
+		idx += n
+	}
+	//TODO Handle Section.TableID=255 as stuffing bytes, but for now this works
+	//Stuff the rest with 0xff
+	for ; idx < len(b); idx++ {
+		b[idx] = 0xff
+	}
+	return idx, nil
+}
+
+func (s *PSISection) Serialise(b []byte) (int, error) {
+
+	if s.Header.TableID == 255 {
+		return 0, nil
+	}
+	idx := 3 // Skip 3 byte header we put in afterward
+
+	if s.Syntax != nil {
+		n, err := s.Syntax.Serialise(b[idx:])
+		if err != nil {
+			return idx, err
+		}
+		idx += n
+	}
+	fmt.Printf("Header length: %x, idx+4: %x idx: %d\n", s.Header.SectionLength, uint16(idx+4), idx)
+
+	s.Header.SectionLength = uint16(idx + 4 - 3) // Add CRC32 field subtract initial 3 bytes
+
+	//Serialise header afterward so we ensure the section length is accurate
+	if s.Header != nil {
+		_, err := s.Header.Serialise(b[0:])
+		if err != nil {
+			return idx, err
+		}
+	}
+
+	if hasCRC32(s.Header.TableType) {
+
+		i := astikit.NewBytesIterator(b)
+		// Get CRC32 data
+		var crc32Data []byte
+		var err error
+		if crc32Data, err = i.NextBytes(idx); err != nil {
+			return idx, fmt.Errorf("astits: fetching next bytes failed: %w", err)
+		}
+
+		// Compute CRC32
+		var crc32 uint32
+		if crc32, err = computeCRC32(crc32Data); err != nil {
+			return idx, fmt.Errorf("astits: computing CRC32 failed: %w", err)
+		}
+
+		// Check CRC32
+		// TODO emit a warning here if it is not valid
+		// if crc32 != s.CRC32 {
+		// 	return idx, fmt.Errorf("astits: Table CRC32 %x != computed CRC32 %x", s.CRC32, crc32)
+		// }
+		b[idx] = uint8(crc32 >> 24)
+		b[idx+1] = uint8(crc32 >> 16)
+		b[idx+2] = uint8(crc32 >> 8)
+		b[idx+3] = uint8(crc32)
+		idx += 4
+	}
+
+	return idx, nil
+
+}
+
+func (h *PSISectionHeader) Serialise(b []byte) (int, error) {
+	// TableID                int    // Table Identifier, that defines the structure of the syntax section and other contained data. As an exception, if this is the byte that immediately follow previous table section and is set to 0xFF, then it indicates that the repeat of table section end here and the rest of TS data payload shall be stuffed with 0xFF. Consequently the value 0xFF shall not be used for the Table Identifier.
+	if h.TableID == 255 {
```

Comment:
Based on the comment above, is this an error?

Reply:
Not technically - it is a stuffing packet if it is 0xFF, so is technically a valid packet just needs to be all bytes set to 0xFF.